File: go/60.go
/*

https://leetcode.com/problems/permutation-sequence/

The set [1,2,3,...,n] contains a total of n! unique permutations.

By listing and labeling all of the permutations in order, we get the following sequence for n = 3:

"123"
"132"
"213"
"231"
"312"
"321"
Given n and k, return the kth permutation sequence.

Note:

Given n will be between 1 and 9 inclusive.
Given k will be between 1 and n! inclusive.
Example 1:

Input: n = 3, k = 3
Output: "213"
Example 2:

Input: n = 4, k = 9
Output: "2314"

https://leetcode.com/problems/permutation-sequence/discuss/22512/Share-my-Python-solution-with-detailed-explanation

*/


package main


import "fmt"

import "strconv"


func factorial(n int) int {
    r := 1
    for n > 0 {
        r = r * n
        n--
    }
    return r
}


func divmod(m int, n int) (int, int) {
    a := m / n
    b := m % n
    return a, b
}


func getPermutation(n int, k int) string {
    nums := make([]string, n)
    for i := 0; i < n; i++ {
        nums[i] = strconv.Itoa(i+1)
    }
    res := ""
    k--
    for n > 0 {
        n--
        i, j := divmod(k, factorial(n))
        res = res + nums[i]
        nums = append(nums[:i], nums[i+1:]...)
        k = j
    }
    return res
}


// 这是一种特定的算法 对于这种特定的算法 只能理解+记忆


func main() {
    fmt.Println(getPermutation(3, 3))
    fmt.Println(getPermutation(3, 5))
    fmt.Println(getPermutation(4, 9))
}

File: python/60.py
"""
https://leetcode.com/problems/permutation-sequence/

The set [1,2,3,...,n] contains a total of n! unique permutations.

By listing and labeling all of the permutations in order, we get the following sequence for n = 3:

"123"
"132"
"213"
"231"
"312"
"321"
Given n and k, return the kth permutation sequence.

Note:

Given n will be between 1 and 9 inclusive.
Given k will be between 1 and n! inclusive.
Example 1:

Input: n = 3, k = 3
Output: "213"
Example 2:

Input: n = 4, k = 9
Output: "2314"

https://leetcode.com/problems/permutation-sequence/discuss/22512/Share-my-Python-solution-with-detailed-explanation

"""


class Solution:
    def getPermutation(self, n: int, k: int) -> str:
        import math
        nums = [str(i) for i in range(1, n + 1)]
        res = ''
        k -= 1
        while n > 0:
            n -= 1
            i, k = divmod(k, math.factorial(n))
            res += nums[i]
            nums.pop(i)

        return res


# 这样的解法


# slow but work

from itertools import permutations

class Solution:
    def getPermutation(self, n: int, k: int) -> str:
        gen = permutations(list(range(1,n+1)))
        r = None
        for _ in range(k):
            r = next(gen)
        return ''.join([str(i) for i in r])


if __name__ == '__main__':
    print(Solution().getPermutation(3, 3))
    print(Solution().getPermutation(3, 5))
    print(Solution().getPermutation(4, 1))
